### PR TITLE
Json object improvements.

### PR DESCRIPTION
**Pull request by rec**
Reviewers: @scottschurr @nbougalis 

If this code looks familiar, it's because it's most of https://github.com/ripple/rippled/pull/832 and has already been reviewed and approved as sch!

Unfortunately, that branch includes a transaction change so it sits around.  I'm extracting the non-breaking changes into a separate change so I can get 'em out...

### CODE REVIEW

**Comment by rec** [comment]
I don't see any way to do that while still preserving the forward declaration...?

**Comment by rec** [comment]
(I did move the two big functions into the .cpp.  The remaining ones will almost certainly want to be inlined anyway as they are near trivial...)

**Comment by scottschurr** [comment]
Tom, since you seem to spend some time in the `*ToJson` stuff, what would you think of turning the `LEDGER_JSON_*` #defines in Ledger.h into an `enum`?  Not necessary for this pull request.  Just a tidy thing to do at some point.  Thanks.

**Comment by rec** [comment]
`LEDGER_JSON_` - wince, this was on my list.  OK, on it right now (in another window!)

**Comment by rec** [comment]
(i.e., I didn't write that but I knew it was there... :-))

**Comment by scottschurr** [comment]
Interesting.  Line 176 of JsonObject.h may be the first case I've ever seen of `protected` inheritance.  Does the fact that `Array` uses private inheritance (line 235) suggest that `Array` should be marked `final`?

**Comment by rec** [comment]
Done, in https://github.com/ripple/rippled/pull/891.

**Comment by scottschurr** [comment]
Looks good to me.
:+1:

**Comment by rec** [comment]
Fixed the spelling mistake and the include guards too.

**Comment by rec** [comment]
The protected inheritance was a suggestion of Howard's, since I never used the base class.  I don't see any particular reason to prevent inheritance, barring any negative consequence from it.   But perhaps we should be marking classes as final by default and changing them if needs be? 

Anyway, I didn't change that.  :-)

**Comment by scottschurr** [comment]
Regarding `final`, my inclination has been to mark anything that wasn't specifically intended to be inherited from as `final`.  And my usual test for whether the class was designed for inheritance is if inheritance actually happens.  That's a luxury we can take since we're not developing library code.  I think the advantage of this conservative approach is, if someone does choose to inherit from a class, they'll know right away that there might be some work to do.  But it's not a big deal.  It's just an inclination I have.

**Comment by vinniefalco** [comment]
@HowardHinnant Which standard library classes are marked `final`?

**Comment by HowardHinnant** [comment]
I just did a search in the latest C++17 draft.  I'm counting zero.

**Comment by nbougalis** [comment]
I didn't notice I was tagged as a reviewer on this - will get on it right now.

On the topic of `final`: I personally tend to think of it as nothing more than window dressing; it's unclear to me that it provides any real, tangible benefit, unlike, e.g. `override`. I'm not going to complain if we add it judiciously when encountering cases during work on business logic, but at the same time, I don't want us to go out of our way to do so.

**Comment by vinniefalco** [comment]
`final` is like making sure the deck chairs are lined up on the Titanic

**Comment by rec** [comment]
Ping!

Do remember that this already got approved as part of https://github.com/ripple/rippled/pull/832...

**Comment by vinniefalco** [comment]
Needs to be squashed down

**Comment by rec** [comment]
The first three changes are essentially separate.  Unfortunately, patching the last change back would involve major surgery due to the move that happened...

**Comment by rec** [comment]
(Next time I'll do the move first...)

**Comment by vinniefalco** [comment]
What's wrong with squashing them all down into one commit with this message?
```
Refactor treament of JSON:
* Better integration between JsonObject and Json::Value
* Move streaming Json objects to ripple/json
* Replace Json::JsonException with std::logic_error
* Move functions to Object.cpp; fix include guards.
```

File: Builds/VisualStudio2013/RippleD.vcxproj [hunk 1]
```diff
@@ -2829,26 +2849,23 @@
     <ClCompile Include="..\..\src\ripple\rpc\tests\Coroutine.test.cpp">
       <ExcludedFromBuild>True</ExcludedFromBuild>
     </ClCompile>
-    <ClCompile Include="..\..\src\ripple\rpc\tests\JsonObject.test.cpp">
-      <ExcludedFromBuild>True</ExcludedFromBuild>
-    </ClCompile>
     <ClCompile Include="..\..\src\ripple\rpc\tests\JSONRPC.test.cpp">
       <ExcludedFromBuild>True</ExcludedFromBuild>
     </ClCompile>
+<<<<<<< HEAD
```
Comment (vinniefalco) [inline Builds/VisualStudio2013/RippleD.vcxproj]: Conflict resolution markers...need to re-run `scons vcxproj`

**Comment by rec** [comment]
Fixed the vcxproj!

I don't want to squash the huge move - which doesn't change behavior - with the change before it or the change after it.  I did squash away one change in there... :-)

**Comment by vinniefalco** [comment]
@rec Fine.
:+1:

**Pull request closed without merging**